File: binary-tree.py
class Node:
    def __init__(self,value):
        self.value = value
        self.left = None
        self.right = None
    
    def insert(self,value):
        if value < self.value: # If value is less than root value go left
            if self.left is None:  # If no left value than set it
                self.left = Node(value)  
            else:                       
                self.left.insert(value) # Repeat process until reached end of tree
        else:                       # If value is more than root value go right
            if self.right is None:   # If no right value than set it
                self.right = Node(value)  # Repeat process until reached end of tree
            else:
                self.right.insert(value)
    
    def in_order_search(self,value): # Search from left to right
        if value < self.value:  # If value is less than root value go left  
            if self.left is None: # If no more values, it was not found
                return f'{value} was not found'
            return self.left.pre_order_search(value)   # Repeat process until either found or reached end of tree
        elif value == self.value:  # If value is found return it
                return f'{value} was found'
        else:                        # If value is more than root value go right
            if self.right is None:   # If no more values, it was not found
                return f'{value} was not found'
            return self.right.pre_order_search(value)  # Repeat process until either found or reached end of tree
    
    def in_order_print(self):  
        if self.left:
            self.left.in_order_print()
        print(self.value)
        if self.right:
            self.right.in_order_print()

    def pre_order_search(self,value): # Search by checking root, then left, then right
        if self.value == value:  # If value is found return it
            return f'{value} was found'
        elif value < self.value:  # If value is less than root value go left                
            if self.left is None:   # If no more values, it was not found
                return f'{value} was not found'
            return self.left.pre_order_search(value)  # Repeat process until either found or reached end of tree
        else:
            if self.right is None:  # If value is more than root value go right
                return f'{value} was not found'  # If no more values, it was not found
            return self.right.pre_order_search(value) # Repeat process until either found or reached end of tree
    
    def pre_order_print(self):
        print(self.value)
        if self.left:
            self.left.pre_order_print()
        if self.right:
            self.right.pre_order_print()

    def post_order_search(self,value):  # Search from left then right then root
        if value < self.value:         # If value is less than root value go left
            if self.left is None:      # If no more values, it was not found
                return f'{value} was not found'  
            return self.left.in_order_search(value)  # Repeat process until either found or reached end of tree
        elif value > self.value:       # If value is more than root value go right
            if self.right is None:      # If no more value, it was not fond
                return f'{value} was not found'
            return self.right.in_order_search(value)   # Repeat process until either found or reached end of tree
        else: 
            return f'{value} was found'
    
    def post_order_print(self):
        if self.left:
            self.left.post_order_print()
        if self.right:
            self.right.post_order_print()
        print(self.value)
    


root = Node(10)
root.insert(7)
root.insert(9)
root.insert(5)
root.insert(14)
root.insert(12)
root.insert(9)
print(root.in_order_search(9))

File: javascript/binaryTree.js
class Node {
    constructor(value) {
        this.left = null;
        this.value = value;
        this.right = null;
    }

    insert(value) {
        if (value < this.value) {
            if (!this.left) this.left = new Node(value);
            else this.left.insert(value)
        }
        else {
            if (!this.right) this.right = new Node(value);
            else this.right.insert(value)
        }
    }

    inOrderSearch(value) {
        if (value < this.value) {
            if (this.left)  this.left.inOrderSearch(value)
            else return false
        }
        else if (value === this.value) return this;
        else {
            if (this.right) this.right.inOrderSearch(value)
            else return false
        }
    }

    inOrderPrint() {
        if (this.left) this.left.inOrderPrint();
        console.log(this.value);
        if (this.right) this.right.inOrderPrint();
    }

    preOrderPrint() {
        console.log(this.value);
        if (this.left) this.left.preOrderPrint()
        if (this.right) this.right.preOrderPrint()
    }

    postOrderPrint() {
        if (this.left) this.left.postOrderPrint()
        if (this.right) this.right.postOrderPrint()
        console.log(this.value)
    }

    invert() {
        if (this.left) this.left.invert();
        if (this.right) this.right.invert();
        const leftVal = this.left;
        this.left = this.right;
        this.right = leftVal;

    }
}

const node = new Node(5);
node.insert(3)
node.insert(7)
node.insert(1)
node.insert(0)
node.insert(2)
node.insert(10)
node.insert(9)
node.insert(12)
node.invert()
node.inOrderPrint()